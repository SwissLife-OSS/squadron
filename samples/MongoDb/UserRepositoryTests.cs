using System.Collections.Generic;
using System.IO;
using System.Threading.Tasks;
using FluentAssertions;
using MongoDB.Driver;
using MongoDB.Driver.Linq;
using Snapshooter.Xunit;
using Squadron.Samples.Shared;
using Xunit;

namespace Squadron.Samples.Mongo
{
    public class UserRepositoryTests : IClassFixture<MongoResource> 
    {
        private readonly MongoResource _mongoResource;

        public UserRepositoryTests(MongoResource mongoResource)
        {
            _mongoResource = mongoResource;
        }

        [Fact]
        public async Task Add_AddedUserIsEquivalent()
        {
            //arrange
            var user = User.CreateSample();
            IMongoDatabase db = _mongoResource.CreateDatabase();
            var repo = new UserRepository(db);

            //act
            await repo.AddAsync(user);

            //assert
            User createdUser = await GetUserAsync(db, user.Id);
            createdUser.Should().BeEquivalentTo(user);
        }

        [Fact]
        public async Task GetAll_MatchSnapshot()
        {
            //arrange
            IMongoDatabase db = _mongoResource.CreateDatabase();
            var options = new CreateCollectionFromFileOptions
            {
                CollectionOptions = new CreateCollectionOptions
                {
                    CollectionName = "users"
                },
                File = new FileInfo("users.json")
            };

            IMongoCollection<User> col = await _mongoResource  
                .CreateCollectionFromFileAsync<User>(db, options);

            var repo = new UserRepository(db);

            //act
            IEnumerable<User> allUsers = await repo.GetAllAsync();

            //assert
            allUsers.MatchSnapshot();
        }

        private Task<User> GetUserAsync(IMongoDatabase db, string id)
        {
            IMongoCollection<User> col = db.GetCollection<User>("users");
            Task<User> user = col.AsQueryable()
                .Where(x => x.Id == id)
                .FirstOrDefaultAsync();
            return user;
        }
    }
}
