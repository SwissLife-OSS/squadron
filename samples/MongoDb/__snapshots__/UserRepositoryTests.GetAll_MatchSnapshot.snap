[
  {
    "Id": "B2",
    "Name": "Sam",
    "Email": "sam.sample@squadron.io"
  },
  {
    "Id": "A1",
    "Name": "John",
    "Email": "john.sample@squadron.io"
  }
]
